token/program-2022/src/extension/mod.rs

```diff
@@ -209,41 +204,65 @@ pub struct StateWithExtensionsMut<'data, S: BaseState> {
     tlv_data: &'data mut [u8],
 }
 impl<'data, S: BaseState> StateWithExtensionsMut<'data, S> {
-    /// Unpack the base state portion of the buffer, leaving the extension data as
-    /// a serialized slice.
+    /// Unpack base state, leaving the extension data as a mutable slice
+    ///
+    /// Fails if the base state is not initialized.
     pub fn unpack(input: &'data mut [u8]) -> Result<Self, ProgramError> {
-        check_not_multisig(input)?;
+        check_min_len_and_not_multisig(input, S::LEN)?;
         let (base_data, rest) = input.split_at_mut(S::LEN);
         let base = S::unpack(base_data)?;
-        let (account_type_index, tlv_start_index) = type_and_tlv_indices::<S>(rest)?;
-        let account_type = AccountType::try_from(rest[account_type_index])
-            .map_err(|_| ProgramError::InvalidAccountData)?;
-        check_account_type::<S>(account_type)?;
-        let (account_type, tlv_data) = rest.split_at_mut(tlv_start_index);
-        Ok(Self {
-            base,
-            base_data,
-            account_type: &mut account_type[account_type_index..tlv_start_index],
-            tlv_data,
-        })
+        if let Some((account_type_index, tlv_start_index)) = type_and_tlv_indices::<S>(rest)? {
+            let account_type = AccountType::try_from(rest[account_type_index])
+                .map_err(|_| ProgramError::InvalidAccountData)?;
+            check_account_type::<S>(account_type)?;
+            let (account_type, tlv_data) = rest.split_at_mut(tlv_start_index);
+            Ok(Self {
+                base,
+                base_data,
+                account_type: &mut account_type[account_type_index..tlv_start_index],
+                tlv_data,
+            })
+        } else {
+            Ok(Self {
+                base,
+                base_data,
+                account_type: &mut [],
+                tlv_data: &mut [],
+            })
+        }
     }
 
-    /// Unpack the base state portion of the buffer without checking for initialization,
-    /// leaving the extension data as a serialized slice.
+    /// Unpack an uninitialized base state, leaving the extension data as a mutable slice
     ///
-    /// The base state of the struct may be totally unusable.
-    pub fn unpack_unchecked(input: &'data mut [u8]) -> Result<Self, ProgramError> {
-        check_not_multisig(input)?;
+    /// Fails if the base state has already been initialized.
+    pub fn unpack_uninitialized(input: &'data mut [u8]) -> Result<Self, ProgramError> {
```

Comment:
"unchecked" is not really what we want IMO.  The calling code really just wants to unpack an account/mint that has yet to be initialized, so make this more explicit

Reply:
You're absolutely right, and a much better name for this.  As I was doing some other work, I kept adding checks in here